hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreFileInfo.java

```diff
@@ -333,76 +331,76 @@ ReaderContext createReaderContext(boolean doDropBehind, long readahead, ReaderTy
       FileNotFoundException exToThrow = null;
       for (int i = 0; i < this.link.getLocations().length; i++) {
         try {
-          return computeHDFSBlocksDistributionInternal(fs);
-        } catch (FileNotFoundException ex) {
-          // try the other location
-          exToThrow = ex;
+          return computeHDFSBlocksDistributionInternal();
+        } catch (FileNotFoundException fnfe) {
+          // Try the other locations -- file behind link may have moved.
+          exToThrow = fnfe;
         }
       }
-      throw exToThrow;
+      throw decorateFileNotFoundException(exToThrow);
     } else {
-      return computeHDFSBlocksDistributionInternal(fs);
+      return computeHDFSBlocksDistributionInternal();
     }
   }
 
-  private HDFSBlocksDistribution computeHDFSBlocksDistributionInternal(final FileSystem fs)
-      throws IOException {
-    FileStatus status = getReferencedFileStatus(fs);
+  private HDFSBlocksDistribution computeHDFSBlocksDistributionInternal() throws IOException {
+    FileStatus status = getFileStatus();
     if (this.reference != null) {
-      return computeRefFileHDFSBlockDistribution(fs, reference, status);
+      return computeRefFileHDFSBlockDistribution(status);
     } else {
-      return FSUtils.computeHDFSBlocksDistribution(fs, status, 0, status.getLen());
+      return FSUtils.computeHDFSBlocksDistribution(this.fs, status, 0, status.getLen());
     }
   }
 
   /**
-   * Get the {@link FileStatus} of the file referenced by this StoreFileInfo
-   * @param fs The current file system to use.
+   * Get the {@link FileStatus} of the file referenced by this StoreFileInfo on the passed
+   * <code>fs</code>.
+   * This {@link StoreFileInfo} could be for a link or a reference or a plain hfile/storefile; get
+   * the filestatus for whatever the link or reference points to (or just the plain hfile/storefile
+   * if not a link/reference). This method is called by snapshot. Filesystem may not be same as
+   * that of this hosting {@link StoreFileInfo}. If a link, when you go to use the passed
+   * FileStatus, the file may have moved; e.g. from data to archive... be aware.
+   * @param fs The filesystem to use.
    * @return The {@link FileStatus} of the file referenced by this StoreFileInfo
    */
   public FileStatus getReferencedFileStatus(final FileSystem fs) throws IOException {
+    if (this.fs.equals(fs) && this.cachedFileStatus != null) {
+      return this.cachedFileStatus;
+    }
     FileStatus status;
     if (this.reference != null) {
       if (this.link != null) {
-        FileNotFoundException exToThrow = null;
-        for (int i = 0; i < this.link.getLocations().length; i++) {
-          // HFileLink Reference
-          try {
-            return link.getFileStatus(fs);
-          } catch (FileNotFoundException ex) {
-            // try the other location
-            exToThrow = ex;
-          }
-        }
-        throw exToThrow;
+        status = this.link.getFileStatus(this.fs);
```

Comment:
The file status to be fetched from this.fs?  Should be from passed param 'fs' ?
And all subsequent calls?   When it is calls within this class, anyways we pass this.fs as param only.  So that part will be ok.

Reply:
Have not checked fully when it calls from Snapshot area.  Just asking Stack.

Reply:
To be safe, it should use the passed param. This method is used outside of this class. I looked at shutting it down but that would be bigger change. Let me update. Thanks @anoopsjohn

Reply:
I closed down the access on this method. Changed snapshot so it creates a StoreFileInfo of its own if different filesystem. Cleaner. Thanks @anoopsjohn